rangefeed: don't delete txn records from the rngfeed pusher

andreimatei opened the pull request:
Before this patch, the rangefeed txn pushed was GC'ing the records of
the transactions that it found to be aborted or committed when it
PUSH_TOUCHED's them. Deleting the txn record is not a nice thing to do
if the txn's coordinator might still be running, and so this patch
terminates the practice. If the coordinator is still running and it had
STAGED the txn record, a PUSH_TOUCH might cause a transition to ABORTED
or COMMITTED. If such a transition is followed by the GC of the txn
record, this causes an ambigous commit error for the coordinator, which
can't tell whether the transaction committed (implicitly) or was rolled
back.

As the code stands, this patch doesn't actually help eliminate the
ambiguity in the ABORTED case (it does in the COMMIT case though),
because the coordinator doesn't distinguish between an ABORTED record
and a missing record; it treats both the same, as ambiguous. But we're
going to improve that, and the coordinator is gonna consider an existing
record as non-ambiguous. The GC'ed case is going to remain the ambiguous
one, and the idea is to avoid that as much as possible by only doing GC
from the gc-queue (after an hour) and from the coordinator.

Touches #52566

Release note (bug fix): Some rare AmbiguousCommitErrors happening when
CDC was used were eliminated.

andreimatei committed:
rangefeed: don't delete txn records from the rngfeed pusher

Before this patch, the rangefeed txn pushed was GC'ing the records of
the transactions that it found to be aborted or committed when it
PUSH_TOUCHED's them. Deleting the txn record is not a nice thing to do
if the txn's coordinator might still be running, and so this patch
terminates the practice. If the coordinator is still running and it had
STAGED the txn record, a PUSH_TOUCH might cause a transition to ABORTED
or COMMITTED. If such a transition is followed by the GC of the txn
record, this causes an ambigous commit error for the coordinator, which
can't tell whether the transaction committed (implicitly) or was rolled
back.

As the code stands, this patch doesn't actually help eliminate the
ambiguity in the ABORTED case (it does in the COMMIT case though),
because the coordinator doesn't distinguish between an ABORTED record
and a missing record; it treats both the same, as ambiguous. But we're
going to improve that, and the coordinator is gonna consider an existing
record as non-ambiguous. The GC'ed case is going to remain the ambiguous
one, and the idea is to avoid that as much as possible by only doing GC
from the gc-queue (after an hour) and from the coordinator.

Touches #52566

Release note (bug fix): Some rare AmbiguousCommitErrors happening when
CDC was used were eliminated.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53146)
<!-- Reviewable:end -->

andreimatei commented:
Only the last commit.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice. That was a pretty clean change.

> Reviewed 1 of 7 files at r5, 6 of 6 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-:-MFCMWmjEr4C_OZAWmQQ:b-8ykqfm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_rangefeed.go, line 107 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCKbUX5Y9iauFM6lAk:-MFCKbUX5Y9iauFM6lAl:b-4tinhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/replica_rangefeed.go#L107)):*
> ```Go
> 	ctx context.Context, intents []roachpb.LockUpdate,
> ) error {
> 	return tp.ir.ResolveIntents(ctx, intents, intentresolver.ResolveOptions{}).GoError()
> ```

Don't we need to poison? That seems pretty important.
___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 549 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCKn2N1YMFz-iu_2RC:-MFCKn2O2G3MfQiq-n1F:b-pngp1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/intentresolver/intent_resolver.go#L549)):*
> ```Go
> // actors to GC a txn record, since that can cause ambiguities for the
> // coordinator: if it had STAGED the txn, it won't be able to tell the
> // difference between a txn that had been implicitly committed and one that
> ```

"implicitly committed, recovered, and GC-ed"

"someone else aborted and GC-ed"
___
*[pkg/kv/kvserver/rangefeed/processor_test.go, line 717 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCLfdjChPkiWMrgoQo:-MFCLfdjChPkiWMrgoQp:b-7iwyh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/processor_test.go#L717)):*
> ```Go
> 	tp.mockResolveIntentsFn(func(ctx context.Context, intents []roachpb.LockUpdate) error {
> 		// There's nothing to assert here. We expect the intents to correspond to
> 		// transactions that had theit LockSpans populated when we pushed them. This
> ```

their
___
*[pkg/kv/kvserver/rangefeed/task.go, line 206 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCM7r4C6KF1oEJm0Xb:-MFCM7r5C7eD-RXdzMmp:bgqdng0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task.go#L206)):*
> ```Go
> 			})
> 
> 			// Asynchronously clean up the transaction's intents, which should
> ```

~Asynchronously~
___
*[pkg/kv/kvserver/rangefeed/task.go, line 231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCM2e7DTzHQzPWD5f4:-MFCM2e7DTzHQzPWD5f5:b31ocmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task.go#L231)):*
> ```Go
> 			// populated. If, however, we ran into a transaction that its coordinator
> 			// tried to rollback but didn't follow up with garbage collection, then
> 			// LockSpans will be populated.
> ```

Lease some indication that all of this is opportunistic and not necessary for correctness.
___
*[pkg/kv/kvserver/rangefeed/task_test.go, line 272 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCMQ_74rm6NIA20v-g:-MFCMQ_74rm6NIA20v-h:bkiu1uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task_test.go#L272)):*
> ```Go
> }
> 
> func (tp *testTxnPusher) intentsToTxns(intents []roachpb.LockUpdate) []enginepb.TxnMeta {
> ```

Does this need to be a method? It doesn't look like it.
___
*[pkg/kv/kvserver/rangefeed/task_test.go, line 286 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCMFUNBABCD44F6nya:-MFCMFUNBABCD44F6nyb:b-77hfpg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task_test.go#L286)):*
> ```Go
> }
> 
> func (tp *testTxnPusher) mockResolveIntentsFn(
> ```

nit: move this up so that the two mocking functions are next to each other.


<!-- Sent from Reviewable.io -->

andreimatei committed:
roachpb: clarify ResolveIntent.Poison

Specify that this field is ignored when resolving COMMITTED (or
otherwise non-ABORTED) intents (which it is). At lease the
rangefeedTxnPusher sets it for non-aborted transactions, leading to
question about whether that's a good idea.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53146)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_rangefeed.go, line 107 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCKbUX5Y9iauFM6lAk:-MFCXxA-AIOJJyk8t5qy:bn9rezh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/replica_rangefeed.go#L107)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Don't we need to poison? That seems pretty important.
</blockquote></details>

good catch; we need to poison on aborted transactions.
Btw, I've prepended a commit commenting on the Poison flag on the proto specifying that it's ignored for intents from non-ABORTED transactions.
___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 549 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCKn2N1YMFz-iu_2RC:-MFCZMhv3ubnY65wPIFY:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/intentresolver/intent_resolver.go#L549)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"implicitly committed, recovered, and GC-ed"

"someone else aborted and GC-ed"
</blockquote></details>

done
___
*[pkg/kv/kvserver/rangefeed/processor_test.go, line 717 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCLfdjChPkiWMrgoQo:-MFCZR6y1mwAmzEd9IhD:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/processor_test.go#L717)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

their
</blockquote></details>

done
___
*[pkg/kv/kvserver/rangefeed/task.go, line 206 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCM7r4C6KF1oEJm0Xb:-MFCZVMF4q0u-h1KzWLg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task.go#L206)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

~Asynchronously~
</blockquote></details>

done
___
*[pkg/kv/kvserver/rangefeed/task.go, line 231 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCM2e7DTzHQzPWD5f4:-MFCZknC7akUjPNWAAMC:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task.go#L231)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Lease some indication that all of this is opportunistic and not necessary for correctness.
</blockquote></details>

done
___
*[pkg/kv/kvserver/rangefeed/task_test.go, line 272 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCMQ_74rm6NIA20v-g:-MFCZt6v8QNlknQxIQr8:bge28mv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task_test.go#L272)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this need to be a method? It doesn't look like it.
</blockquote></details>

it doesn't but I'd rather namespace it here
___
*[pkg/kv/kvserver/rangefeed/task_test.go, line 286 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCMFUNBABCD44F6nya:-MFC_CV58zBFVDc5jt3z:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c62f6a5ba12e71c8a3a2665a1ea8fd0d759516e/pkg/kv/kvserver/rangefeed/task_test.go#L286)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move this up so that the two mocking functions are next to each other.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 8 files at r7, 6 of 6 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-:-MFCmNq6COP7j99owqs-:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 785 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCm5nuAO9E1gJmvCxo:-MFCm5nuAO9E1gJmvCxp:b-y1kon9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/kv/kvserver/intentresolver/intent_resolver.go#L785)):*
> ```
> arent'
> ```

aren't
___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 786 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCm95JBHLVJ0OdmyDM:-MFCm95JBHLVJ0OdmyDN:bixkict) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/kv/kvserver/intentresolver/intent_resolver.go#L786)):*
> ```Go
> 	//
> 	// This field is ignored for intents that arent' resolved for an ABORTED txn;
> 	// in other words only intents from ABORTED transactions ever poison the abort
> ```

Missing comma after "words"
___
*[pkg/roachpb/api.proto, line 1111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCmKFaDFway5SkdWxV:-MFCmKFaDFway5SkdWxW:b-yfs00f) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/roachpb/api.proto#L1111)):*
> ```protobuf
>   TransactionStatus status = 3;
>   // Optionally poison the abort span for the transaction on all ranges
>   // on which the intents reside. The field is ignored if status != ABORTED (i.e. only intents from
> ```

Looks like the wrapping got messed up here.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-:-MFCw9q43ZMRpDt6VIRY:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 785 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCm5nuAO9E1gJmvCxo:-MFCuyfp0GODhh_n_sd_:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/kv/kvserver/intentresolver/intent_resolver.go#L785)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> arent'
> ```

aren't
</blockquote></details>

done
___
*[pkg/kv/kvserver/intentresolver/intent_resolver.go, line 786 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCm95JBHLVJ0OdmyDM:-MFCv1KE6rVxsdIb_MUB:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/kv/kvserver/intentresolver/intent_resolver.go#L786)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Missing comma after "words"
</blockquote></details>

done
___
*[pkg/roachpb/api.proto, line 1111 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/53146#-MFCmKFaDFway5SkdWxV:-MFCvM6n8xCAldsivOKs:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9742b8cd10c9a4c195634be7877b0b78eb61/pkg/roachpb/api.proto#L1111)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Looks like the wrapping got messed up here.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->
